codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -121,35 +124,91 @@ public void close() { }
     @Override
     public ChannelFuture writeData(ChannelHandlerContext ctx, int streamId, ByteBuf data,
             int padding, boolean endStream, ChannelPromise promise) {
-        boolean releaseData = true;
+        boolean lastHeaderWritten = false;
+        boolean lastFrameWritten = false;
         SimpleChannelPromiseAggregator promiseAggregator =
                 new SimpleChannelPromiseAggregator(promise, ctx.channel(), ctx.executor());
+        ByteBuf headerBuf = null;
         try {
             verifyStreamId(streamId, STREAM_ID);
             verifyPadding(padding);
 
-            Http2Flags flags = new Http2Flags().paddingPresent(padding > 0).endOfStream(endStream);
+            // Determine the size of the padding presence field, if padding is present.
+            final int paddingPresenceFieldLength = padding > 0 ? 1 : 0;
 
-            int payloadLength = data.readableBytes() + padding + flags.getPaddingPresenceFieldLength();
-            verifyPayloadLength(payloadLength);
+            int remainingData = data.readableBytes();
 
-            ByteBuf buf = ctx.alloc().buffer(DATA_FRAME_HEADER_LENGTH);
-            writeFrameHeaderInternal(buf, payloadLength, DATA, flags, streamId);
-            writePaddingLength(buf, padding);
-            ctx.write(buf, promiseAggregator.newPromise());
+            // Determine the number of frames that we're about to send. Padding will be split
+            // evenly across the frames, so we subtract off the size of the padding presence field
+            // from each frame (if padding is present).
+            final int numFrames = fastCeil((remainingData + padding) /
```

Comment:
We have had an issue open on padding for a while https://github.com/netty/netty/issues/2807. I also just assume put the same padding on each frame unless we have some rational for doing something more complicated.